### PR TITLE
Fix invalid handling of null DateTime parameters in Npgsql 6+

### LINKED ISSUE #3291: Npgsql 6+ issues with null DateTime parameter types
Hi!
I am sorry, it seems theese changes leads to an error in some cases.

Versions:
- 	NHibernate 5.4.2
- 	Npgsql 7.0.2

The named query:
```
	<sql-query name="MyQuery" cache-mode="normal" read-only="true">
	  <query-param type="DateTime" name="DateOfSearch"/>
	  <return-scalar type="Int64" column="SomeID"/>
	  <![CDATA[select ...
	        from ...
	        where ((  :DateOfSearch is null ) or (( t.DT1 <= :DateOfSearch) and ( t.DT2 >= :DateOfSearch )))
	        and ... ;]]>
	</sql-query>
	</hibernate-mapping>
```
Call the query:
```
	DateTime? dateOfSearch = null; // set null value here -- !!!!
	query.SetParameter("DateOfSearch", dateOfSearch, NHibernateUtil.DateTime);
	var result = query.List<long>();
```
NHibernate log:
```
	2023-04-21 23:01:51,433 [1] DEBUG NHibernate.Type.DateTimeType - binding null to parameter: 0
	2023-04-21 23:01:51,433 [1] DEBUG NHibernate.Type.DateTimeType - binding null to parameter: 1
	2023-04-21 23:01:51,434 [1] DEBUG NHibernate.Type.DateTimeType - binding null to parameter: 2
	2023-04-21 23:01:51,445 [1] DEBUG NHibernate.SQL - select ...
	        from ...
	        where ((  :p0 is null ) or (( mmh.DT_Mount <= :p0) and ( mmh.DT_Dismount >= :p0 ))); :p0 = NULL [Type: Object (0:0:0)]  -- !!!!
	...
	2023-04-21 23:01:51,623 [1] ERROR NHibernate.AdoNet.AbstractBatcher - Could not execute query: select ...
```
`dbParam.DBType` is empty, and Npgsql sends `Prepare` command to PostgreSQL with no parameter type specified, but PostgreSQL also can not guess the type from this query and value.
It works correctly, if I set non-null parameter value.

_Originally posted by @olga-sonix in https://github.com/nhibernate/nhibernate-core/pull/3064#discussion_r1174135624_

### PR DESCRIPTION
**Pull request by hazzik**
Fixes #3291

### CODE REVIEW

**Review by fredericDelaporte — APPROVED** [review]

**Comment by fredericDelaporte** [comment]
I have thought about checking where the regression had happened only after validating. It seems it was in 5.3.x. So we need need a backport of the fix, now that it is merged into 5.4.

**Review by bahusoid — COMMENTED** [review]
> It seems it was in 5.3.x

Yes, original fix was added in 5.3.13. But it’s not a regression as DateTime handling was broken anyway with Npgsql6 (see #2994). So I would make it a low priority 5.3.x fix that can wait for other regression issues.

File: src/NHibernate/Driver/NpgsqlDriver.cs [hunk 1]
```diff
@@ -74,14 +75,28 @@ protected override void InitializeParameter(DbParameter dbParam, string name, Sq
...
+
+		public override void AdjustCommand(DbCommand command)
+		{
+			if (DriverVersionMajor >= 6)
 			{
-				// Let Npgsql 6 driver to decide parameter type 
+				for (var i = 0; i < command.Parameters.Count; i++)
+				{
+					var parameter = command.Parameters[i];
+					if (parameter.Value is DateTime)
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/NpgsqlDriver.cs:91]: > It's not clear what original issue #2994 is about as no details are provided.

There are breaking changes in Npgsql 6 that require a DateTime of certain Kind to match timestamp or timestamptz through a "help" of DbType.DateTime2 type. Here are some explanations: https://www.npgsql.org/doc/release-notes/6.0.html#detailed-notes and here: https://www.roji.org/postgresql-dotnet-timestamp-mapping

It is really hard to explain their motivation and the implementation. But it does not align with the type system used by NHibernate. 

> But DateTime can represent multiple DbType types (Date, DateTime, Time).

I don't think it applies here. However, I think, it is possible to write a more specific conversion.

[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/NpgsqlDriver.cs:91]: Ok, it does apply to DbType.Date. I'll adjust the fix.

[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Driver/NpgsqlDriver.cs:91]: @bahusoid https://github.com/nhibernate/nhibernate-core/pull/3301

**Pull request merged**